File: mailblastr-ruby/lib/mailblastr/contact_properties.rb
# frozen_string_literal: true

module Mailblastr
  # Custom contact fields, usable as {{merge_tags}}.
  module ContactProperties
    class << self
      # POST /contact-properties — params: { key:, type: "string"|"number", fallback_value: }
      def create(params)
        Client.request(:post, "/contact-properties", body: params)
      end

      # GET /contact-properties/:id
      def get(property_id)
        Client.request(:get, "/contact-properties/#{Client.path_escape(property_id)}")
      end

      # GET /contact-properties
      def list(params = {})
        Client.request(:get, "/contact-properties", query: Client.pagination(params))
      end

      # PATCH /contact-properties/:id — only fallback_value is mutable.
      def update(property_id, params)
        Client.request(:patch, "/contact-properties/#{Client.path_escape(property_id)}", body: params)
      end

      # DELETE /contact-properties/:id
      def delete(property_id)
        Client.request(:delete, "/contact-properties/#{Client.path_escape(property_id)}")
      end
    end
  end
end

File: mailblastr-php/src/Resources/ContactProperties.php
<?php

declare(strict_types=1);

namespace Mailblastr\Resources;

use Mailblastr\Client;

/**
 * Contact properties (custom fields usable as {{merge_tags}}).
 */
class ContactProperties extends Resource
{
    /**
     * Register a contact property. POST /contact-properties
     * Returns the slim ack ['object' => 'contact_property', 'id' => …].
     *
     * @param array $payload ['key' => … ('name' accepted as an alias),
     *                       'type' => 'string'|'number', 'fallback_value' => …]
     */
    public function create(array $payload): array
    {
        return $this->client->request('POST', '/contact-properties', $payload);
    }

    /** Retrieve a contact property. GET /contact-properties/:id */
    public function get(string $id): array
    {
        return $this->client->request('GET', '/contact-properties/' . Client::e($id));
    }

    /**
     * List contact properties. GET /contact-properties
     *
     * @param array $params Cursor pagination: limit, after, before.
     */
    public function list(array $params = []): array
    {
        return $this->client->request('GET', '/contact-properties' . $this->paginationQuery($params));
    }

    /**
     * Update a contact property (only fallback_value is mutable; key/type are
     * immutable). PATCH /contact-properties/:id
     */
    public function update(string $id, array $payload): array
    {
        return $this->client->request('PATCH', '/contact-properties/' . Client::e($id), $payload);
    }

    /** Delete a contact property. DELETE /contact-properties/:id */
    public function remove(string $id): array
    {
        return $this->client->request('DELETE', '/contact-properties/' . Client::e($id));
    }
}
